geometry/meshop.cpp

```diff
@@ -252,6 +252,118 @@ void saveAsPly( const Mesh &mesh, const boost::filesystem::path &filepath){
     }
 }
 
+void saveAsPlyWithFeatures(const boost::filesystem::path& filepath,
+                           const Mesh& mesh,
+                           const math::Points3i& vertexColors,
+                           const math::Points3i& faceColors,
+                           const math::Points3& vertexNormals,
+                           const std::vector<int>& faceLabels)
+{
+    LOG(info2) << "Saving mesh to file <" << filepath << ">.";
+
+    std::ofstream out(filepath.string().c_str());
+    out.setf(std::ios::scientific, std::ios::floatfield);
+
+    unsigned validFaces(0);
+    for (const auto& face : mesh.faces)
+    {
+        if (!face.degenerate() && mesh.good(face)) validFaces++;
+    }
+
+    // create header
+    out << "ply\n"
+        << "format ascii 1.0\n"
+        << "comment generated by window-mesh\n"
+        << "element vertex " << mesh.vertices.size() << '\n'
+        << "property float x\n"
+        << "property float y\n"
+        << "property float z\n";
+
+    if (vertexColors.size())
+    {
+        utility::expect(vertexColors.size() == mesh.vertices.size(),
+                        "Length of vertexColors must be the same as number of "
+                        "mesh vertices.");
+        out << "property uchar red\n"
+            << "property uchar green\n"
+            << "property uchar blue\n";
+    }
+
+    if (vertexNormals.size())
+    {
+        utility::expect(vertexNormals.size() == mesh.vertices.size(),
+                        "Length of vertexNormals must be the same as number of "
+                        "mesh vertices.");
+        out << "property float nx\n"
+            << "property float ny\n"
+            << "property float nz\n";
+    }
+
+    out << "element face " << validFaces << '\n'
+        << "property list uchar int vertex_indices\n";
+
+    if (faceColors.size())
+    {
+        utility::expect(faceColors.size() == validFaces,
```

Comment:
It's a little bit strange to expect the `faceColor` vector to be already prefiltered while the `mesh.faces` are not. Bc now everyone who wants to use this function has to create the faceColor vector in the very same way as is done the filtering here (line 270).

Reply:
True, I followed `loadPly()` too blindly in this respect. I removed the check for degenerate faces altogether.